background/get-userscripts.js

```diff
@@ -17,61 +17,104 @@ function getL10NURLs() {
   return urls;
 }
 
+scratchAddons.localEvents.addEventListener("addonEnabled", ({ detail }) => {
+  const { addonId, manifest } = detail;
+  chrome.tabs.query({}, (tabs) =>
+    tabs.forEach((tab) => {
+      if (tab.url || (!tab.url && typeof browser !== "undefined")) {
+        chrome.tabs.sendMessage(tab.id, "getInitialUrl", { frameId: 0 }, async (res) => {
+          if (res) {
+            const { userscripts, userstyles } = await getAddonData({ addonId, url: res, manifest });
+
+            console.log(userscripts, userstyles);
+            chrome.tabs.sendMessage(
+              tab.id,
+              {
+                dynamicAddonEnabled: {
+                  scripts: userscripts,
+                  userstyles,
+                  addonId,
+                  injectAsStyleElt: manifest.injectAsStyleElt,
+                  index: scratchAddons.manifests.findIndex((addon) => addon.addonId === addonId),
+                },
+              },
+              { frameId: 0 }
+            );
+          }
+        });
+      }
+    })
+  );
+});
+scratchAddons.localEvents.addEventListener("addonDisable", ({ detail }) => {
+  const { addonId, manifest } = detail;
+  chrome.tabs.query({}, (tabs) =>
+    tabs.forEach((tab) => {
+      if (tab.url || (!tab.url && typeof browser !== "undefined")) {
+        chrome.tabs.sendMessage(tab.id, "getInitialUrl", { frameId: 0 }, async (res) => {
+          if (res) {
+            chrome.tabs.sendMessage(tab.id, { dynamicAddonDisable: { addonId } }, { frameId: 0 });
+          }
+        });
+      }
+    })
+  );
+});
+
+async function getAddonData({ addonId, manifest, url }) {
+  const promices = [];
+
+  const userscripts = [];
+  for (const script of manifest.userscripts || []) {
+    if (userscriptMatches({ url }, script, addonId))
+      userscripts.push({
+        url: script.url,
+        runAtComplete: typeof script.runAtComplete === "boolean" ? script.runAtComplete : true,
+      });
+  }
+  const userstyles = [];
+  for (const style of manifest.userstyles || []) {
+    if (userscriptMatches({ url }, style, addonId))
+      if (manifest.injectAsStyleElt) {
+        promices.push(
+          fetch(chrome.runtime.getURL(`/addons/${addonId}/${style.url}`))
+            .then((res) => res.text())
+            .then((text) => {
+              userstyles.push(text);
+            })
```

Comment:
This pattern might cause race conditions, see #700

Reply:
Not so sure... this is the userstyles between an addon... I don't think an addon's own files should conflict. But if they do, we can add a simple sort function based on how the manifest lists the userstyles.

Reply:
It's also worth noting that this is how it worked before this PR...

Reply:
> We can add a simple sort function based on how the manifest lists the userstyles.

It's as simple as reserving all needed indices of the array, then replacing the correct item depending in the order established in the manifest, instead of pushing to the end.

> It's also worth noting that this is how it worked before this PR...

No. As of now, theme userstyles (injected as style elements) are provided through a `styleUrls` array that is properly ordered, and a `styles` property is used to map URLs to the actual contents of the style element. See #700 for more info.